### PR TITLE
03 array paulfrmbrn

### PR DESCRIPTION
**Pull request by PaulFrmBrn**
Here is the implementation. I found few bugs in tests and have one suggestion though.

### CODE REVIEW

File: src/main/java/array/Array.java [hunk 1]
```diff
@@ -40,4 +40,11 @@
      * @param coordinates - position for deleting
      */
     void delete(C coordinates);
+
+    /**
+     * Visit the array (print for example)
+     *
+     * @param consumer - array's cell visitor
+     */
+    void visit(Consumer<T> consumer);
```
Comment (PaulFrmBrn) [inline src/main/java/array/Array.java:49]: There were no such method in here and signature of the method for visiting 1d and 2d array was different* SO I've got a suggestion to introduce common visit method. Please check the code below

File: src/main/java/array/Array2d.java [hunk 2]
```diff
@@ -19,35 +36,73 @@
...
     }
 
-    /**
-     * Visit the array (print for example)
-     *
-     * @param rowConsumer - array's row visitor
-     */
-    public void visit(Consumer<Array1d<T>> rowConsumer) {
-        throw new UnsupportedOperationException();
+    public void visit(Consumer<T> consumer) {
```
Comment (PaulFrmBrn) [inline src/main/java/array/Array2d.java:82]: This is what I'm talking about. 
In current implementation `Array2d` knows about `Array1d`. I've missed that while review, sorry.

File: src/test/java/array/Array1dTest.java [hunk 3]
```diff
@@ -99,7 +103,7 @@ void shouldVisitNonEmptyArray() {
                 .forEach(i -> array.insert(
                         Coordinates1d.of(i),
                         10 * i));
-        String expectedPresentation = "0, 1, 2, 3, 4, 5, 6, 7, 8, null ";
```
Comment (PaulFrmBrn) [inline src/test/java/array/Array1dTest.java:102]: It seems there is a mistake here

File: src/test/java/array/Array2dTest.java [hunk 4]
```diff
@@ -98,17 +103,14 @@ void shouldVisitNonEmptyArray() {
                 .forEach(i -> array.insert(
                         Coordinates2d.of(i, i),
                         10 * i));
-        String expectedPresentation = "0, null, null, null, null, \n" +
-                                      "null, 1, null, null, null, \n" +
```
Comment (PaulFrmBrn) [inline src/test/java/array/Array2dTest.java:102]: It seems there is a mistake here

File: src/main/java/array/Array.java (hunk 1 above)
Comment (PaulFrmBrn) [reply src/main/java/array/Array.java]: separate PR

File: src/main/java/array/Array2d.java (hunk 2 above)
Comment (PaulFrmBrn) [reply src/main/java/array/Array2d.java:82]: separate PR

File: src/test/java/array/Array1dTest.java (hunk 3 above)
Comment (PaulFrmBrn) [reply src/test/java/array/Array1dTest.java]: separate PR

File: src/test/java/array/Array2dTest.java (hunk 4 above)
Comment (PaulFrmBrn) [reply src/test/java/array/Array2dTest.java]: separate PR

**Comment by zamonier** [comment]
Wrong target branch

File: src/main/java/array/Array2d.java [hunk 5]
```diff
@@ -19,31 +36,73 @@
...
+        }
     }
 
     @Override
     public Optional<Coordinates2d> search(T value) {
-        throw new UnsupportedOperationException();
+        validateValue(value);
+        for (int i = 0; i < numberOfRows; i++) {
+            for (int j = 0; j < numberOfColumns; j++) {
+                if (array[i][j] != null && array[i][j].equals(value)) {
```
Comment (zamonier) [inline src/main/java/array/Array2d.java:68]:
```suggestion
                if (value.equals(array[i][j])) {
```
'cause value is already not null

File: src/main/java/recursion/Factorial.java [hunk 6]
```diff
@@ -3,6 +3,12 @@
 public class Factorial {
```
Comment (zamonier) [inline src/main/java/recursion/Factorial.java:3]: ?

File: src/main/java/recursion/Fibonacci.java [hunk 7]
```diff
@@ -3,6 +3,15 @@
 public class Fibonacci {
```
Comment (zamonier) [inline src/main/java/recursion/Fibonacci.java:3]: ?

File: src/main/java/array/Array1d.java [hunk 8]
```diff
@@ -18,32 +27,67 @@
...
+            return Optional.of(value);
+        }
     }
 
     @Override
     public Optional<Coordinates1d> search(T value) {
-        throw new UnsupportedOperationException();
+        validateValue(value);
+        for (int i = 0; i < array.length; i++) {
+            if (array[i] != null && array[i].equals(value)) {
```
[the remark this answers is not in the document]
Comment (PaulFrmBrn) [reply src/main/java/array/Array1d.java]: fixed

File: src/main/java/array/Array2d.java [hunk 9]
```diff
@@ -19,31 +36,73 @@
...
+        array[coordinates.getRow()][coordinates.getColumn()] = value;
+        return value;
     }
 
     @Override
     public Optional<T> at(Coordinates2d coordinates) {
-        throw new UnsupportedOperationException();
+        validateIndex(coordinates);
+        T value = (T) array[coordinates.getRow()][coordinates.getColumn()];
+        if (value == null) {
```
[the remark this answers is not in the document]
Comment (PaulFrmBrn) [reply src/main/java/array/Array2d.java]: fixed

File: src/main/java/array/Array2d.java (hunk 5 above)
Comment (PaulFrmBrn) [reply src/main/java/array/Array2d.java]: fixed

File: src/main/java/array/Array1d.java [hunk 10]
```diff
@@ -18,32 +27,67 @@
...
+        array[coordinates.getIndex()] = value;
+        return value;
     }
 
     @Override
     public Optional<T> at(Coordinates1d coordinates) {
-        throw new UnsupportedOperationException();
+        validateIndex(coordinates);
+        T value = (T) array[coordinates.getIndex()];
+        if (value == null) {
```
[the remark this answers is not in the document]
Comment (PaulFrmBrn) [reply src/main/java/array/Array1d.java]: fixed

File: src/main/java/recursion/Factorial.java (hunk 6 above)
Comment (PaulFrmBrn) [reply src/main/java/recursion/Factorial.java]: fixed

File: src/main/java/recursion/Fibonacci.java (hunk 7 above)
Comment (PaulFrmBrn) [reply src/main/java/recursion/Fibonacci.java]: fixed

**Pull request merged**